src/core/thread/mle_router.cpp

```diff
@@ -1670,8 +1670,17 @@ ThreadError MleRouter::HandleParentRequest(const Message &aMessage, const Ip6::M
     SuccessOrExit(error = Tlv::GetTlv(aMessage, Tlv::kChallenge, sizeof(challenge), challenge));
     VerifyOrExit(challenge.IsValid(), error = kThreadError_Parse);
 
-    if ((child = FindChild(macAddr)) == NULL)
+    child = FindChild(macAddr);
+
+    if (child != NULL && !(child->mMode & ModeTlv::kModeFFD))
+    {
+        // Parent Request from a MTD child means that the child had detached. It can be safely removed.
+        RemoveNeighbor(*child);
+    }
+
+    if (child == NULL || !(child->mMode & ModeTlv::kModeFFD))
```

Comment:
This will work given the existing code.  However, I think it would be safer to set `child` to `NULL` just after `RemoveNeighbor()` above in case `RemoveNeighbor()` has other side effects in the future.  Then we can remove the additional `mMode` check here.